### PR TITLE
Update mypy and improve config

### PR DESCRIPTION
**Pull request by ethanhs**
This has some new nice things like `TypeGuard`

### CODE REVIEW

**Comment by ethanhs** [comment]
FWIW, the needed changes have been made upstream. I'm going to add type ignores to silence these errors for now.

**Comment by edyounis** [comment]
Can we change the upstream branch to iterative-reoptimization since that significantly changes the partitioners, which is very relevant to these errors? Otherwise LGTM

File: tests/conftest.py [hunk 1]
```diff
@@ -97,16 +97,20 @@
     'float_9': np.longdouble(1234.0),
     'float_10': np.float32(1234.0),
     'float_11': np.float64(1234.0),
-    'complex_1': complex(0.0j),
-    'complex_2': complex(0.0 + 0.0j),
+    # Needs typeshed sync
```
Comment (edyounis) [inline tests/conftest.py:100]: All this should go soon when we start working with python hypothesis.

**Review by edyounis — APPROVED** [review]
LGTM, thanks for the update!

**Review by edyounis — APPROVED** [review]
LGTM, thanks for the update!

**Pull request merged**